federation-js/src/service/printCoreSchema.ts

```diff
@@ -0,0 +1,581 @@
+}
+
+function printScalar(type: GraphQLScalarType, options?: Options): string {
+  return printDescription(options, type) + `scalar ${type.name}`;
+}
+
+function printObject(
+  type: GraphQLObjectType,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const interfaces = type.getInterfaces();
+  const implementedInterfaces = interfaces.length
+    ? ' implements ' + interfaces.map((i) => i.name).join(' & ')
+    : '';
+
+  return (
+    printDescription(options, type) +
+    `type ${type.name}` +
+    implementedInterfaces +
+    // Core addition for printing @join__owner and @join__type usages
+    printTypeJoinDirectives(type, context) +
+    printFields(options, type, context)
+  );
+}
+
+// Core change: print @join__owner and @join__type usages
+function printTypeJoinDirectives(
+  type: GraphQLObjectType,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const metadata: FederationType = type.extensions?.federation;
+  if (!metadata) return '';
+
+  const { serviceName: ownerService, keys } = metadata;
+  if (!ownerService || !keys) return '';
+
+  // Separate owner @keys from the rest of the @keys so we can print them
+  // adjacent to the @owner directive.
+  const { [ownerService]: ownerKeys = [], ...restKeys } = keys;
+  const ownerEntry: [string, (readonly SelectionNode[])[]] = [
+    ownerService,
+    ownerKeys,
+  ];
+  const restEntries = Object.entries(restKeys);
+
+  return (
+    `\n  @join__owner(graph: ${
+      context?.sanitizedServiceNames[ownerService] ?? ownerService
+    })` +
+    [ownerEntry, ...restEntries]
+      .map(([service, keys = []]) =>
+        keys
+          .map(
+            (selections) =>
+              `\n  @join__type(graph: ${
+                context?.sanitizedServiceNames[service] ?? service
+              }, key: ${printStringLiteral(printFieldSet(selections))})`,
+          )
+          .join(''),
+      )
+      .join('')
+  );
+}
+
+function printInterface(
+  type: GraphQLInterfaceType,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  return (
+    printDescription(options, type) +
+    `interface ${type.name}` +
+    printFields(options, type, context)
+  );
+}
+
+function printUnion(type: GraphQLUnionType, options?: Options): string {
+  const types = type.getTypes();
+  const possibleTypes = types.length ? ' = ' + types.join(' | ') : '';
+  return printDescription(options, type) + 'union ' + type.name + possibleTypes;
+}
+
+function printEnum(type: GraphQLEnumType, options?: Options): string {
+  const values = type
+    .getValues()
+    .map(
+      (value, i) =>
+        printDescription(options, value, '  ', !i) +
+        '  ' +
+        value.name +
+        printDeprecated(value) +
+        printDirectivesOnEnumValue(type, value),
```

Comment:
I should have a comment here like `// Core change: ...` - this looks like a standard printer fn but it's not. I should probably rename this to `maybePrintJoinGraphEnumValueDirectives` or something.